frame/election-provider-multi-phase/src/lib.rs

```diff
@@ -0,0 +1,1495 @@
+/// [`Pallet::feasibility_check`]
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug)]
+pub struct RawSolution<C> {
+	/// Compact election edges.
+	compact: C,
+	/// The _claimed_ score of the solution.
+	score: ElectionScore,
+	/// The round at which this solution should be submitted.
+	round: u32,
+}
+
+impl<C: Default> Default for RawSolution<C> {
+	fn default() -> Self {
+		// Round 0 is always invalid, only set this to 1.
+		Self { round: 1, compact: Default::default(), score: Default::default() }
+	}
+}
+
+/// A checked solution, ready to be enacted.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, Default)]
+pub struct ReadySolution<A> {
+	/// The final supports of the solution.
+	///
+	/// This is target-major vector, storing each winners, total backing, and each individual
+	/// backer.
+	supports: Supports<A>,
+	/// The score of the solution.
+	///
+	/// This is needed to potentially challenge the solution.
+	score: ElectionScore,
+	/// How this election was computed.
+	compute: ElectionCompute,
+}
+
+/// Solution size of the election.
+///
+/// This is needed for proper weight calculation.
+#[derive(PartialEq, Eq, Clone, Copy, Encode, Decode, RuntimeDebug, Default)]
+pub struct SolutionSize {
+	/// Number of all voters.
+	///
+	/// This must match the on-chain snapshot.
+	#[codec(compact)]
+	voters: u32,
+	/// Number of all targets.
+	///
+	/// This must match the on-chain snapshot.
+	#[codec(compact)]
+	targets: u32,
+}
+
+/// A snapshot of all the data that is needed for en entire round. They are provided by
+/// [`ElectionDataProvider`] at the beginning of the signed phase (or the unsigned phase, if signed
+/// phase is non-existent) and are kept around until the round is finished.
+///
+/// These are stored together because they are often times accessed together.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, Default)]
+pub struct RoundSnapshot<A> {
+	/// All of the voters.
+	pub voters: Vec<(A, VoteWeight, Vec<A>)>,
+	/// All of the targets.
+	pub targets: Vec<A>,
+}
+
+/// Some metadata related to snapshot.
+///
+/// In this pallet, there are cases where we want to read the whole snapshot (voters, targets,
+/// desired), and cases that we are interested in just the length of these values. The former favors
+/// the snapshot to be stored in one struct (as it is now) while the latter prefers them to be
+/// separate to enable the use of `decode_len`. This approach is a middle ground, storing the
+/// snapshot as one struct, whilst storing the lengths separately.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, Default)]
+pub struct RoundSnapshotMetadata {
+	/// The length of voters.
+	voters_len: u32,
+	/// The length of targets.
+	targets_len: u32,
+}
+
+/// Internal errors of the pallet.
+///
+/// Note that this is different from [`pallet::Error`].
+#[derive(RuntimeDebug, Eq, PartialEq)]
+pub enum ElectionError {
```

Comment:
Need to check, but I guess this is a bit messy and contains errors from different processes. Ideally, there should be at least a separate one for miner operations, `enum MinerError`, some variants could wrap other stuff.

Reply:
Cleaned it up, and even added a document section to the module doc about how the errors are laid out. A bit verbose, but intentional and happy with the outcome.